openedx_learning/core/components/api.py

```diff
@@ -18,32 +18,44 @@
 from django.db.models import Q, QuerySet
 from django.db.transaction import atomic
 
+from ...lib.cache import lru_cache
 from ..publishing import api as publishing_api
-from .models import Component, ComponentVersion, ComponentVersionRawContent
+from .models import Component, ComponentType, ComponentVersion, ComponentVersionRawContent
+
+
+@lru_cache(maxsize=128)
+def get_or_create_component_type_id(namespace: str, name: str) -> int:
```

Comment:
Even though it returns a pk, I think I would mildly prefer `get_or_create_component_type`, since you're not going to create a "compoment type", not a "component type id".

Reply:
Ah. Would the rolled-back id be mis-cached in the `lru_cache`, or somewhere else?

Reply:
Yeah, in `lru_cache`.

Reply:
I don't expect that _creating_ `ComponentType` instances will occur outside of `create_component`. So I suggest that you just put the `ComponentType.objects.get_or_create(` call inline (without a cache) inside `create_component`. Then if it gets rolled back, no problem.

If this gets simplified to `@lru_cache def get_component_type_id()` then you have two nice simplifications:
1. If the `ComponentType` is not found, it won't be cached, due to the `ComponentType.NotFound` exception
2. You can optimize lookups like this:
    ```python
     Component.with_publishing_relations \
                        .get(
                            learning_package_id=learning_package_id,
                            component_type__namespace=namespace,
                            component_type__name=type_name,
                            local_key=local_key,
                        )
    ```

    becomes

    ```python
     Component.with_publishing_relations \
                        .get(
                            learning_package_id=learning_package_id,
                            component_type_id=get_component_type_id(namespace, type_name),
                            local_key=local_key,
                        )
    ```
which will then cause your function to throw differentiated exceptions: `Component.NotFound` OR `ComponentType.NotFound` vs. only throwing `Component.NotFound`. Arguably more explicit.

Reply:
^ Nice, I like that.

Reply:
I think that I'm going to just kick the `lru` to the curb entirely for now. The worry I have is that the `lru` bleeds across transactions, and we don't know what combination of reads and writes of Components people are going to do in a big transaction (e.g. import). So it's possible that a process will add a bunch of stuff (including new ContentTypes), then do some reads of Components because they're grouping them somehow (thus loading things into the LRU cache). Then their process dies, and gets rolled back.

I guess the semantics I'm really looking for are transaction-local caches. Which actually sounds really cool now that I think about it... but I'm definitely not going there this week. A much simpler version of that would be a local object that proxies the requests through and goes out of scope at the end of whatever import operation happens. But things will probably be fine if I just punt on caching for a while.